src/error.rs

```diff
@@ -1,16 +1,158 @@
-//! Error chain
-use error_chain::*;
-pub(crate) use error_chain::bail;
+//! Error handling
+//!
+//! A module allowing us to create more detailed, context aware, errors.
+//!
+//! # Examples
+//!
+//! Invalid UTC argument type:
+//!
+//! ```rust
+//! use balena_temen::{
+//!     ast::Identifier,
+//!     Engine, Context, Value,
+//!     error::*
+//! };
+//! use std::collections::HashMap;
+//!
+//! let engine: Engine = Engine::default();
+//! let mut ctx = Context::default();
+//! let position = Identifier::default();
+//! let data = Value::Null;
+//!
+//! eprintln!("{}", engine.eval("now(utc=`yes`)", &position, &data, &mut ctx).err().unwrap());
+//! ```
+//!
+//! Printed error:
+//!
+//! ```text
+//! temen: invalid argument type
+//!  └ context:
+//!     ├ function = now
+//!     ├ argument name = utc
+//!     ├ argument value = "yes"
+//!     └ expected = boolean
+//! ```
+use std::borrow::Cow;
+use std::error;
+use std::fmt;
+use std::result;
 
-use crate::parser::Rule;
+/// Standard library result wrapper
+pub type Result<T> = result::Result<T, Error>;
 
-error_chain! {
-    errors {}
+/// Context key type
+pub type Key = Cow<'static, str>;
+/// Context value type
+pub type Value = Cow<'static, str>;
 
-    foreign_links {
-        Json(serde_json::Error) #[doc = "JSON (de)serialization error"];
-        Pest(pest::error::Error<Rule>) #[doc = "Parser error"];
-        ParseIntError(std::num::ParseIntError) #[doc = "Integer parser error"];
-        ParseFloatError(std::num::ParseFloatError) #[doc = "Float parser error"];
+/// Result extension
+pub trait ResultExt<T> {
+    /// Appends key, value pair to the context
+    ///
+    /// # Arguments
+    ///
+    /// * `k` - A key
+    /// * `v` - A value
+    fn context<K, V>(self, k: K, v: V) -> Result<T>
+    where
+        K: Into<Key>,
+        V: Into<Value>;
+}
+
+impl<T> ResultExt<T> for Result<T> {
+    fn context<K, V>(self, k: K, v: V) -> Result<T>
+    where
+        K: Into<Key>,
+        V: Into<Value>,
+    {
+        self.map_err(|e| e.context(k, v))
+    }
+}
+
+/// Error type
+pub struct Error {
+    inner: Box<Inner>,
```

Comment:
Is box needed because we expect the error to be rather big ?
If so - the comment could help readers here, to list examples in which box could be beneficial ?
If not - we could potentially get rid of `Inner` and just have `Error` with all the fields ? What do you think ? (I removed the Box to try and everything seems happy)

Reply:
Did add comment about Box. The reason is to keep `Result` as small as possible.